File: src/client/java/fr/hardel/asset_editor/client/javafx/components/layout/SupportCard.java
package fr.hardel.asset_editor.client.javafx.components.layout;

import fr.hardel.asset_editor.client.javafx.VoxelResourceLoader;
import fr.hardel.asset_editor.client.javafx.VoxelFonts;
import fr.hardel.asset_editor.client.javafx.components.ui.Button;
import fr.hardel.asset_editor.client.javafx.components.ui.SvgIcon;
import fr.hardel.asset_editor.client.javafx.lib.utils.BrowserUtils;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.control.Label;
import javafx.scene.effect.ColorAdjust;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.ColumnConstraints;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Priority;
import javafx.scene.layout.Region;
import javafx.scene.layout.StackPane;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;
import javafx.scene.shape.Rectangle;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.resources.Identifier;

public final class SupportCard extends StackPane {

    private static final Identifier LOGO = Identifier.fromNamespaceAndPath("asset_editor", "icons/logo.svg");
    private static final Identifier SHINE = Identifier.fromNamespaceAndPath("asset_editor",
        "textures/shine.png");
    private static final Identifier CHECK = Identifier.fromNamespaceAndPath("asset_editor", "icons/check.svg");
    private static final Identifier PATREON_ICON = Identifier.fromNamespaceAndPath("asset_editor",
        "icons/company/patreon.svg");
    private static final String[] ADVANTAGES = { "early_access", "submit_ideas", "discord_role", "live_voxel" };

    public SupportCard() {
        getStyleClass().add("layout-support-card");
        setAlignment(Pos.TOP_LEFT);

        SvgIcon logo = new SvgIcon(LOGO, 384, Color.WHITE);
        logo.setOpacity(0.2);
        logo.setManaged(false);
        logo.setMouseTransparent(true);

        Label title = new Label(I18n.get("supports:title"));
        title.getStyleClass().add("layout-support-card-title");
        title.setFont(VoxelFonts.of(VoxelFonts.Variant.SEMI_BOLD, 30));
        title.setWrapText(true);

        Label desc = new Label(I18n.get("supports:description"));
        desc.getStyleClass().add("layout-support-card-desc");
        desc.setFont(VoxelFonts.of(VoxelFonts.Variant.REGULAR, 14));
        desc.setWrapText(true);
        VBox.setMargin(desc, new Insets(8, 0, 0, 0));

        VBox textBlock = new VBox(title, desc);

        Label advantagesHeading = new Label(I18n.get("supports:advantages"));
        advantagesHeading.getStyleClass().add("layout-support-card-advantages-heading");
        advantagesHeading.setFont(VoxelFonts.of(VoxelFonts.Variant.BOLD, 20));
        VBox.setMargin(advantagesHeading, new Insets(24, 0, 16, 0));

        GridPane advantagesGrid = buildAdvantagesGrid();
        VBox advantagesSection = new VBox(advantagesHeading, advantagesGrid);
        HBox.setHgrow(advantagesSection, Priority.ALWAYS);

        Button donateBtn = new Button(Button.Variant.SHIMMER, Button.Size.LG, I18n.get("supports:donate"));
        donateBtn.setOnAction(() -> BrowserUtils.openBrowser("https://streamelements.com/hardoudou/tip"));

        SvgIcon patreonIcon = new SvgIcon(PATREON_ICON, 16, Color.WHITE);
        Button patreonBtn = new Button(Button.Variant.PATREON, Button.Size.LG, I18n.get("supports:become"),
            patreonIcon);
        patreonBtn.setOnAction(() -> BrowserUtils.openBrowser("https://www.patreon.com/hardel"));
        donateBtn.setMaxWidth(Double.MAX_VALUE);
        patreonBtn.setMaxWidth(Double.MAX_VALUE);
        HBox.setHgrow(donateBtn, Priority.ALWAYS);
        HBox.setHgrow(patreonBtn, Priority.ALWAYS);

        HBox buttonsRow = new HBox(16, donateBtn, patreonBtn);
        buttonsRow.setAlignment(Pos.CENTER_RIGHT);
        buttonsRow.setPadding(new Insets(32, 0, 0, 0));

        Region btnSpacer = new Region();
        VBox.setVgrow(btnSpacer, Priority.ALWAYS);
        VBox buttonsWrapper = new VBox(btnSpacer, buttonsRow);
        buttonsWrapper.setAlignment(Pos.BOTTOM_RIGHT);

        HBox bottom = new HBox(16, advantagesSection, buttonsWrapper);
        bottom.setPadding(new Insets(16, 0, 0, 0));

        VBox cardContent = new VBox(textBlock, bottom);
        cardContent.setPadding(new Insets(32, 32, 32, 48));
        StackPane.setAlignment(cardContent, Pos.TOP_LEFT);

        try (var stream = VoxelResourceLoader.open(SHINE)) {
            ImageView shine = new ImageView(new Image(stream));
            ColorAdjust shineEffect = new ColorAdjust();
            shineEffect.setHue(0.5);
            shineEffect.setBrightness(-0.8);
            shine.setEffect(shineEffect);
            shine.setPreserveRatio(false);
            shine.setManaged(false);
            shine.setMouseTransparent(true);
            shine.fitWidthProperty().bind(widthProperty());
            shine.fitHeightProperty().bind(heightProperty());
            getChildren().add(shine);
        } catch (Exception ignored) {
        }

        logo.layoutXProperty().bind(widthProperty().subtract(288));
        logo.setLayoutY(-96);
        getChildren().addAll(logo, cardContent);

        Rectangle clip = new Rectangle();
        clip.setArcWidth(32);
        clip.setArcHeight(32);
        widthProperty().addListener((obs, o, w) -> clip.setWidth(w.doubleValue()));
        heightProperty().addListener((obs, o, h) -> clip.setHeight(h.doubleValue()));
        setClip(clip);
    }

    private static GridPane buildAdvantagesGrid() {
        GridPane grid = new GridPane();
        grid.setHgap(32);
        grid.setVgap(16);
        for (int i = 0; i < 2; i++) {
            ColumnConstraints c = new ColumnConstraints();
            c.setHgrow(Priority.ALWAYS);
            grid.getColumnConstraints().add(c);
        }
        for (int i = 0; i < ADVANTAGES.length; i++) {
            grid.add(buildAdvantageItem("supports:advantages." + ADVANTAGES[i]), i % 2, i / 2);
        }
        return grid;
    }

    private static HBox buildAdvantageItem(String key) {
        SvgIcon check = new SvgIcon(CHECK, 16, Color.WHITE);
        Label text = new Label(I18n.get(key));
        text.getStyleClass().add("layout-support-card-benefit");
        text.setFont(VoxelFonts.of(VoxelFonts.Variant.SEMI_BOLD, 14));
        HBox item = new HBox(8, check, text);
        item.setAlignment(Pos.CENTER_LEFT);
        return item;
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/components/layout/SupportCard.kt
package fr.hardel.asset_editor.client.compose.components.layout

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.unit.dp
import fr.hardel.asset_editor.AssetEditor
import fr.hardel.asset_editor.client.compose.StudioColors
import fr.hardel.asset_editor.client.compose.StudioTypography
import fr.hardel.asset_editor.client.compose.components.ui.Button
import fr.hardel.asset_editor.client.compose.components.ui.ButtonSize
import fr.hardel.asset_editor.client.compose.components.ui.ButtonVariant
import fr.hardel.asset_editor.client.compose.components.ui.ShineOverlay
import fr.hardel.asset_editor.client.compose.components.ui.SvgIcon
import fr.hardel.asset_editor.client.compose.lib.utils.BrowserUtils
import net.minecraft.client.resources.language.I18n
import net.minecraft.resources.Identifier

private val LOGO = Identifier.fromNamespaceAndPath(AssetEditor.MOD_ID, "icons/logo.svg")
private val CHECK = Identifier.fromNamespaceAndPath(AssetEditor.MOD_ID, "icons/check.svg")
private val PATREON_ICON = Identifier.fromNamespaceAndPath(AssetEditor.MOD_ID, "icons/company/patreon.svg")
private val ADVANTAGES = listOf("early_access", "submit_ideas", "discord_role", "live_voxel")
private val cardShape = RoundedCornerShape(16.dp)
private val borderColor = StudioColors.Zinc900

@Composable
fun SupportCard(modifier: Modifier = Modifier) {
    // bg-black/35 border-t-2 border-l-2 rounded-2xl border-stone-900 overflow-hidden relative
    Box(
        modifier = modifier
            .fillMaxWidth()
            .clip(cardShape)
            .background(Color.Black.copy(alpha = 0.35f), cardShape)
            .drawWithContent {
                drawContent()
                val stroke = 2.dp.toPx()
                val hs = stroke / 2f
                val r = 16.dp.toPx()
                val arcSize = Size((r - hs) * 2, (r - hs) * 2)

                drawLine(borderColor, Offset(hs, size.height), Offset(hs, r), strokeWidth = stroke)
                drawArc(borderColor, 180f, 90f, false, Offset(hs, hs), arcSize, style = Stroke(stroke))
                drawLine(borderColor, Offset(r, hs), Offset(size.width, hs), strokeWidth = stroke)
            }
    ) {
        ShineOverlay(modifier = Modifier.matchParentSize(), opacity = 0.15f)
        Box(
            contentAlignment = Alignment.TopEnd,
            modifier = Modifier.matchParentSize()
        ) {
            SvgIcon(
                location = LOGO,
                size = 384.dp,
                tint = Color.White,
                modifier = Modifier
                    .offset(x = 96.dp, y = (-96).dp)
                    .alpha(0.2f)
            )
        }

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(start = 48.dp, end = 32.dp, top = 32.dp, bottom = 32.dp)
        ) {
            Text(
                text = I18n.get("supports:title"),
                style = StudioTypography.semiBold(30),
                color = Color.White
            )
            Text(
                text = I18n.get("supports:description"),
                style = StudioTypography.regular(14),
                color = StudioColors.Zinc400,
                modifier = Modifier.padding(top = 8.dp)
            )

            Row(
                horizontalArrangement = Arrangement.spacedBy(16.dp),
                modifier = Modifier.fillMaxWidth().padding(top = 16.dp)
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = I18n.get("supports:advantages"),
                        style = StudioTypography.bold(20),
                        color = Color.White,
                        modifier = Modifier.padding(top = 24.dp, bottom = 16.dp)
                    )

                    Row(
                        horizontalArrangement = Arrangement.spacedBy(32.dp),
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Column(
                            verticalArrangement = Arrangement.spacedBy(16.dp),
                            modifier = Modifier.weight(1f)
                        ) {
                            ADVANTAGES.filterIndexed { i, _ -> i % 2 == 0 }.forEach {
                                AdvantageItem("supports:advantages.$it")
                            }
                        }
                        Column(
                            verticalArrangement = Arrangement.spacedBy(16.dp),
                            modifier = Modifier.weight(1f)
                        ) {
                            ADVANTAGES.filterIndexed { i, _ -> i % 2 == 1 }.forEach {
                                AdvantageItem("supports:advantages.$it")
                            }
                        }
                    }
                }

                Row(
                    horizontalArrangement = Arrangement.spacedBy(16.dp),
                    verticalAlignment = Alignment.Bottom,
                    modifier = Modifier.align(Alignment.Bottom).padding(top = 32.dp)
                ) {
                    Button(
                        onClick = { BrowserUtils.openBrowser("https://streamelements.com/hardoudou/tip") },
                        variant = ButtonVariant.SHIMMER,
                        size = ButtonSize.LG,
                        text = I18n.get("supports:donate")
                    )
                    Button(
                        onClick = { BrowserUtils.openBrowser("https://www.patreon.com/hardel") },
                        variant = ButtonVariant.PATREON,
                        size = ButtonSize.LG,
                        text = I18n.get("supports:become"),
                        icon = { SvgIcon(PATREON_ICON, 16.dp, Color.White) }
                    )
                }
            }
        }
    }
}

@Composable
private fun AdvantageItem(key: String) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        SvgIcon(CHECK, 16.dp, Color.White)
        Text(
            text = I18n.get(key),
            style = StudioTypography.semiBold(14),
            color = StudioColors.Zinc300
        )
    }
}
